### ISSUE TITLE
AudioPacketType.SequenceEnd is not defined

### ISSUE THREAD

**Issue by BtbN** [veovera/enhanced-rtmp#29]
The spec uses AudioPacketType.SequenceEnd in one place, but it has never defined a value for it anywhere.

**Comment by zenomt** [comment]
agreed that the spec has logic for an audio SequenceEnd in one spot. however, an "audio silence message" (an empty audio message with no codec or frame type) is traditionally used to signal the end of an audio sequence (and for the decoder to flush and for playback to behave as though there's no longer an audio track, until new audio frames arrive).

i think the correct solution should be to just remove the
```
if (audioPacketType == AudioPacketType.SequenceEnd) {
    // signals end of sequence
}
```
from the audio section, and instead document the traditional (but currently undocumented) "audio silence message" and its intended behavior. i think an audio sequence end message would just be more bytes to say the same thing as an audio silence message, and is unnecessary.

**Comment by BtbN** [comment]
It does not increase the size though, just changes half a byte. And it's a bit cleaner imo.
Not like saving bytes matters at that scale. We're talking saving individual bytes here.

**Comment by zenomt** [comment]
using the "audio silence message" actually saves a whole byte, because it's _just_ a message of type "audio" with no payload at all (no codec nibble, no frame type nibble), *and* the silence message and its meaning (at least in Flash Player, AIR, and Adobe Media Server) already exists and should be documented anyway.

having multiple codings for the same semantic message i think is less clean, since there would need to be two playback pathways to achieve the same result.

**Comment by veovera** [comment]
Great catch and an excellent discussion!

Not having a value for the `AudioPacketType.SequenceEnd` enum is an oversight from one of the merges. In fact, it is present in our internal prototype `pseudocode.ts` file; otherwise, VS Code would not be able to compile it. Some thoughts:

- **Correct, we could save a byte by leveraging an audio silence packet instead of `AudioPacketType.SequenceEnd` enum, but one byte savings is not compelling at this scale.**
- **The original intent of the silence packet was optimization.** It was created to save bytes on the wire. For example, if a publishing client was capturing voice audio and noticed the user stopped talking, it could pause sending audio data and instead send a silence packet to indicate silence on the publishing end. The audio messages would resume once the user started talking again. Thus, there could be many silence packets if the solution decided to implement this optimization. The silence packet was not meant to be a `SequenceEnd` to a `SequenceStart`.
- **If we keep `AudioPacketType.SequenceEnd`, we should mention in the spec that there should be no more data to parse within that packet.** Even though the spec mentions that the pseudocode logic is not exhaustive, it might be a good idea to call this out explicitly in the pseudocode or elsewhere in the spec.
- **The currently undocumented silence packet should probably be mentioned in the spec.** @BtbN, does FFmpeg currently have logic for the silence packet when parsing RTMP streams?
- **Keeping `AudioPacketType.SequenceEnd` and documenting the silence packet is what I am leaning towards.**

**Comment by BtbN** [comment]
Right now, whenever FFmpeg flvdec encounters an empty packet of any kind, it'll just move on to the next packet:
https://github.com/FFmpeg/FFmpeg/blob/master/libavformat/flvdec.c#L1507

It does not even check for SequenceEnd or anything. It'll just keep going until the file or stream has come to an end.

**Comment by zenomt** [comment]
the semantics of the "audio silence" message in the Flash Media playback and timing model is:
* flush the audio decoder (if any)
* indication that no more audio might be coming for a while, so:
* switch to wall-clock timing for video + data playback, instead of using the audio message timestamps and the audio sample clock for the master playback clock
* don't wait for audio frames for synchronized A+V playback

other than for uses covered by the above behaviors, in what circumstance would a distinct "audio end-of-sequence" ever be used? Flash Media semantics already provide (also not exactly documented, but Flash Player always worked this way) that you can change codecs on-the-fly, or send a new "sequence start" message, and the reasonable thing when doing so is to flush the previous instance of the decoder and (re)initialize.

**Comment by zenomt** [comment]
"end of sequence" would be useful in multitrack mode, since there's no way to have a _completely empty_ audio message for just one track ID, for cases where one track might be at an end or transition point or discontinuity that's different from other tracks. once you're in multitrack mode and talking about a track, there's a message type and a codec fourCc, so it's not "empty".

**Comment by veovera** [comment]
> where one track might be at an end or transition point or discontinuity that's different from other tracks. once you're in multitrack mode and talking about a track, there's a message type and a codec fourCc, so it's not "empty".

Nice catch!

**Comment by veovera** [comment]
[AudioPacketType.SequenceEnd](https://github.com/veovera/enhanced-rtmp/blob/b843cdd6c314574661baac73f9b29d68432b254a/docs/enhanced/enhanced-rtmp-v2.md?plain=1#L593-L626) is now defined. We have also documented the traditional audio silence packet which was previously undocumented.

**Comment by zenomt** [comment]
> We have also documented the traditional audio silence packet which was previously undocumented.

the wording in the spec for the audio silence message is a little confusing and maybe ambiguous. it says

> This silence message is identified when an audio message contains a zero payload, indicating a period of silence.

i think you want to say "... a zero **length** payload", or more precisely describe an "empty audio message" (that is, not even having an AudioTagHeader). my reading of "zero payload" suggests a payload consisting of bytes with value `0`.

**Comment by veovera** [comment]
> the wording in the spec for the audio silence message is a little confusing

Good point. We will reword it to be:

```
// RTMP includes a previously undocumented 'audio silence' message.
// This silence message is identified when an audio message contains 
// a zero-length payload, or more precisely, an empty audio message 
// without an AudioTagHeader, indicating a period of silence. The 
// action to take after receiving a silence message is system 
// dependent. The semantics of the silence message in the Flash 
// Media playback and timing model are as follows:

```
